flist.c

```diff
@@ -231,10 +232,15 @@ static int readlink_stat(const char *path, STRUCT_STAT *stp, char *linkbuf)
 int link_stat(const char *path, STRUCT_STAT *stp, int follow_dirlinks)
 {
 #ifdef SUPPORT_LINKS
-	if (copy_links)
+	if (copy_links && update_links == 0)
 		return x_stat(path, stp, NULL);
 	if (x_lstat(path, stp, NULL) < 0)
 		return -1;
+	if (update_links && S_ISLNK(stp->st_mode)) {
+		STRUCT_STAT st;
+		if (x_stat(path, &st, NULL) == 0 && !S_ISDIR(st.st_mode))
+			return x_lstat(path, stp, NULL);
+	}
```

Comment:
This hunk baffles me. It does a stat of the referent file, and if it's a non-dir it re-stats the symlink's stat that is already present in stp and returns. But if it is not a dir, the following code also returns the same stat data that is already in stp. So, it seems like this can be dropped as well.

Reply:
Precisely, if the symlink points at a non-directory referent, return the symlink's stat rather than the referent or fall through. It's only apparent looking at the surrounding code why the x_lstat call is redundant. Committed as `aed6235`, works on assumption lstat or lstat64 is guaranteed to return either 0 or -1 on all platforms.